### DISCUSSION TITLE
Problem with overwriting dynamic meta props at page level

### DISCUSSION THREAD

**Discussion by tavyy** [vercel/next.js#43186] [Help] [answered]
So I'm using `@bradgarropy/next-seo` to set my meta props now. I have tried using Next's Head component as well and also got the same problem.

I have the following at _app.js:

```JS
function MyApp({ Component, pageProps }) {
  return (
    <div>
      <ThirdwebProvider desiredChainId={activeChainId}>
        <SEO title="..." description="..." facebook={{
              image: "...",
              url: "....",
              type: "website",
            }} 
        />
        <Component {...pageProps} />
      </ThirdwebProvider>  
    </div>
  )
}
```

This works fine. However, I need to set overwrite the dynamic meta props at the page level. As per documentation, I do:

pages/some_page.js

```JS
if (loadingState === 'not-loaded') {
    return(
      <div>
        <div className={styles.container}>
          <Navigation/>
        </div>
        <div className="flex justify-center items-center mt-7">
          <Image src={loadingImg} className="object-none object-center" />
        </div>
      </div>
    )
  } else if (Object.keys(post).length === 0) {
    return (
      <div>
        <div className={styles.container}>
          <Navigation/>
        </div>

        <h1 className="px-20 py-10 text-3xl">Item not found.</h1>
      </div>
    )
  }

  return (
    <div>
  

      <SEO title="...." description={post.description} facebook={{
              image: post.file,
              url: `...`,
              type: "website",
            }} 
        />
```

When I check the DOM in the console, I can see the meta props. But in the View Source section, I only see the default ones defined at _app.js. When I check with say FB's OG debugger it always renders the props defined at the _app.js level.

It is driving me crazy because no matter what I do I can't overwrite those meta props at the page level. I do have conditions on the page as per code, but that is needed for UX, so that is mandatory.

None of the solutions posted on Github or Stackoverflow work as if those people never bothered to actually test what they are suggesting.

Is Next.JS that poorly designed that it can't achieve this simple thing?

**Comment by icyJoseph** [comment, marked as answer]
> Is Next.JS that poorly designed that it can't achieve this simple thing?

Isn't `next-seo` a third party package? You could just do this manually, but we given in to these packages API to ease our work. Let's be reasonable here and avoid non-contributing adjectives. 

Now let's look at your issue. If you use a REST client, such as Postman, or [VSCode](https://marketplace.visualstudio.com/items?itemName=humao.rest-client), and make a request for your site, you'll see the `_app` meta props, correct? You could also just deactivate JavaScript for your dev page in whichever browser you are using.

Do yo also see the content for this branch `loadingState === 'not-loaded'`?

**Reply by tavyy** [reply]
```JS
Isn't next-seo a third party package? You could just do this manually, but we given in to these packages API to ease our work. Let's be reasonable here and avoid non-contributing adjectives.
```
- I have also used Next's Head component and got the exact same result as my original post states.

```JS
Now let's look at your issue. If you use a REST client, such as Postman, or [VSCode](https://marketplace.visualstudio.com/items?itemName=humao.rest-client), and make a request for your site, you'll see the _app meta props, correct?
``` 
- Correct

```JS
Do yo also see the content for this branch loadingState === 'not-loaded'?
```
- If I do a curl now I don't see this. I see the content that is supposed to show when the page is loaded.

```JS
You could also just deactivate JavaScript for your dev page in whichever browser you are using.
```
- this is not a local dev issue. It happens on prod as well.

**Reply by icyJoseph** [reply]
Ah yeah but the page with JS disabled has to show the correct meta tags. It was just an alternative to using `curl`. 

> If I do a curl now I don't see this. I see the content that is supposed to show when the page is loaded.

Yet, it has incorrect meta tags? I feel like I've fixed this previously, and also whatever I found online didn't fix it.

**Reply by tavyy** [reply]
Yes with incorrect meta tags. It's using the default ones instead of the page-level ones. The funny thing is that if I inspect the page using the dev tools, the correct meta-tags are shown in the DOM tree. But the page source however shows the wrong ones.

**Reply by icyJoseph** [reply]
Aha, I made a mistake, you are using: [@bradgarropy/next-seo](https://github.com/bradgarropy/next-seo), and not [next-seo](https://github.com/garmeeh/next-seo), though you used the regular Head and still saw issues. 🤔 .

**Reply by tavyy** [reply]
Yes. I used the regular Head originally and then decided to use some 3rd party stuff hoping that would maybe fix the problem. Using `"next": "12.2.2"` btw.

**Reply by icyJoseph** [reply]
Ah there was one of my next questions. Alright I'm downgrading to that.

Also, a bit of a strange question, but are all tags missing? or only some?

**Reply by tavyy** [reply]
No, they are not missing they just have the wrong valuer. They all display the default value defined at the _app.js level. And if that was the question then yes they all have the wrong value. No exceptions. As if the overwrite doesn't apply. Also if you test, can you please test with a page that has conditions in it as I do? Some people say that might be the cause, but I can't take out the conditions as it is just going to break everything. I am doing an API call and waiting for the answer for a few ms, so the loaded/not-loaded state is a must.

**Reply by icyJoseph** [reply]
Alright, a create next-app changed to next@12.2.2 with _app:

```jsx
import "../styles/globals.css";
import type { AppProps } from "next/app";
import Head from "next/head";

export default function App({ Component, pageProps }: AppProps) {
  return (
    <div>
      <Head>
        <title>Fallback</title>
      </Head>
      <Component {...pageProps} />
    </div>
  );
}
```

And a home page:

```jsx
import Head from "next/head";

export default function Home() {
  return (
    <div>
      <Head>
        <title>Home title</title>
      </Head>

      <p>home</p>
    </div>
  );
}
```

Is working tip-top, so we ought to suspect something else is going on here. 

Is it possible that by accident, some `Head` tag is wrapping the SEO at some level? I do recall that nested Heads counter each other.

**Reply by tavyy** [reply]
Can you add a condition on that page as I have? Something like this: 

```JS
if (loadingState === 'not-loaded') {
    return(
      <div>
        <div className={styles.container}>
          <Navigation/>
        </div>
        <div className="flex justify-center items-center mt-7">
          <Image src={loadingImg} className="object-none object-center" />
        </div>
      </div>
    )
  }

return (
....
```

And also instead of `title` can you use something like: `<meta property="og:image" content="bla"/>` although I don't think it would make much difference, but to have it as close as mine.

**Reply by icyJoseph** [reply]
> Also if you test, can you please test with a page that has conditions in it as I do? Some people say that might be the cause, but I can't take out the conditions as it is just going to break everything. 

Then though, the HTML content would also be incorrect, because the server wouldn't be touching those paths, and that'd keep the tags. That's why I opened with that line of question.

```tsx
import Head from "next/head";
import { useState } from "react";

export default function Home() {
  const [state, setState] = useState(false);

  if (state) return <div></div>;

  return (
    <div>
      <Head>
        <title>Home title</title>
      </Head>

      <p>home</p>
    </div>
  );
}
```

The above works, but if I start with `true` then as expected, it fails. However, you said that curl returns the correct branch, where the meta is defined.

**Reply by tavyy** [reply]
Yes but I wonder if there is an issue with the DOM being updated during the state transition. Because initially not-loaded state is true and then it becomes false. So in your example, you would have to somehow simulate the state transition from true -> false after a few ms wait time.

**Reply by icyJoseph** [reply]
For meta tags, you gotta add keys though, as per docs https://nextjs.org/docs/api-reference/next/head:

> To avoid duplicate tags in your head you can use the key property, which will make sure the tag is only rendered once, as in the following example:

```js
import Head from 'next/head'

function IndexPage() {
  return (
    <div>
      <Head>
        <title>My page title</title>
        <meta property="og:title" content="My page title" key="title" />
      </Head>
      <Head>
        <meta property="og:title" content="My new title" key="title" />
      </Head>
      <p>Hello world!</p>
    </div>
  )
}
```

> In this case only the second <meta property="og:title" /> is rendered. meta tags with duplicate key attributes are automatically handled.

Also:

> title, meta or any other elements (e.g. script) need to be contained as direct children of the Head element, or wrapped into maximum one level of <React.Fragment> or arrays—otherwise the tags won't be correctly picked up on client-side navigations.

**Reply by tavyy** [reply]
> For meta tags, you gotta add keys though, as per docs https://nextjs.org/docs/api-reference/next/head:
> 
> > To avoid duplicate tags in your head you can use the key property, which will make sure the tag is only rendered once, as in the following example:
> 
> ```js
> import Head from 'next/head'
> 
> function IndexPage() {
>   return (
>     <div>
>       <Head>
>         <title>My page title</title>
>         <meta property="og:title" content="My page title" key="title" />
>       </Head>
>       <Head>
>         <meta property="og:title" content="My new title" key="title" />
>       </Head>
>       <p>Hello world!</p>
>     </div>
>   )
> }
> ```
> 
> > In this case only the second  is rendered. meta tags with duplicate key attributes are automatically handled.
> 
> Also:
> 
> > title, meta or any other elements (e.g. script) need to be contained as direct children of the Head element, or wrapped into maximum one level of <React.Fragment> or arrays—otherwise the tags won't be correctly picked up on client-side navigations.

Yep already familiar with that. And I did that.

**Reply by icyJoseph** [reply]
> Yes but I wonder if there is an issue with the DOM being updated during the state transition. Because initially not-loaded state is true and then it becomes false. So in your example, you would have to somehow simulate state transition from true -> false in a few ms

Could you please try in the browser with JS disabled, what branch of content do you see? In the server there's no transition, the pages directory renders a long HTML string with zero interactions. So if initially, `loadingState === 'not-loaded'` is `true` the server renders that branch, and it never "transitions" into the other branch. Suspense boundaries and React 18 can alter that, but I am assuming that's not what we are dealing with here.

**Reply by icyJoseph** [reply]
Is there a public repo to look at?

**Reply by tavyy** [reply]
> > Yes but I wonder if there is an issue with the DOM being updated during the state transition. Because initially not-loaded state is true and then it becomes false. So in your example, you would have to somehow simulate state transition from true -> false in a few ms
> 
> Could you please try in the browser with JS disabled, what branch of content do you see? In the server there's no transition, the pages directory renders a long HTML string with zero interactions. So if initially, `loadingState === 'not-loaded'` is `true` the server renders that branch, and it never "transitions" into the other branch. Suspense boundaries and React 18 can alter that, but I am assuming that's not what we are dealing with here.

Let me try to do that.

**Reply by tavyy** [reply]
Ok so with JS disabled I just see the loader and that's it which means it gets stuck at:

```JS
if (loadingState === 'not-loaded') {
    return(
      <div>
        <div className={styles.container}>
          <Navigation/>
        </div>
        <div className="flex justify-center items-center mt-7">
          <Image src={loadingImg} className="object-none object-center" />
        </div>
      </div>
    )
  }
```

So no transition of state happens. And the OG meta tags are there but again they have the default values which is expected in this case.

**Reply by tavyy** [reply]
> Is there a public repo to look at?

I mean what time zone are you I could send you a quick invite on a quick zoom session screen sharing and send you a beer if you fix it :D ? Because I already have everything setup. It doesn't have to be today btw.

**Reply by tavyy** [reply]
Oh damn and I forgot to mention I also have `_document.js` that looks like this:

```JS
import Document, { Html, Head, Main, NextScript } from "next/document";

class MainDocument extends Document {
  static async getInitialProps(ctx) {
    const initialProps = await Document.getInitialProps(ctx);
    return { ...initialProps };
  }

  render() {
    return (
      <Html>
        <Head>
       
          {/* Global Site Tag (gtag.js) - Google Analytics */}
          <script
            async
            src={`https://www.googletagmanager.com/gtag/js?id=XXXX`}
          />
          <script
            dangerouslySetInnerHTML={{
              __html: `
            window.dataLayer = window.dataLayer || [];
            function gtag(){dataLayer.push(arguments);}
            gtag('js', new Date());
            gtag('config', 'G-XXXXX', {
              page_path: window.location.pathname,
            });
          `,
            }}

          />

        {/* Hotjar Tracking Code */ }
        <script
          dangerouslySetInnerHTML={{
              __html: `(function(h,o,t,j,a,r){
                h.hj=h.hj||function(){(h.hj.q=h.hj.q||[]).push(arguments)};
                h._hjSettings={hjid:3167289,hjsv:6};
                a=o.getElementsByTagName('head')[0];
                r=o.createElement('script');r.async=1;
                r.src=t+h._hjSettings.hjid+j+h._hjSettings.hjsv;
                a.appendChild(r);
            })(window,document,'https://static.hotjar.com/c/hotjar-','.js?sv=');
            `,
          }}
        />
        
        </Head>
        <body>
          <Main />
          
          <NextScript />
          {/*Below we add the modal wrapper*/}
          <div id="modal-root"></div>
          <script src="https://unpkg.com/@themesberg/flowbite@latest/dist/flowbite.bundle.js" async></script>
          <script type="text/javascript" id="hs-script-loader" async defer src="//js-eu1.hs-scripts.com/26536362.js"></script>
   
        </body>
      </Html>
    );
  }
}

export default MainDocument;
```

**Reply by icyJoseph** [reply]
Yeah so FB, Google and other non JS entities, see the first HTML frame without your meta tags. 

One quick and dirty solution is to fetch the post inside `getServerSideProps`, or if possible pre-render the data, with `getStaticProps`, while still fetching fresh data client side, but with ISR revalidation that step feels a bit unnecessary, in some scenarios.

Basically:

- If you want the latest data at all times, no matter what, `GSSP` is your friend
- If it can be stale first, and then (on refresh or after some time) be revalidated, `GSP` or `ISR`. 


I am a bit intrigued that `curl` did show the UI though, maybe a misunderstanding.

**Reply by tavyy** [reply]
> Yeah so FB, Google and other non JS entities, see the first HTML frame without your meta tags.
> 
> One quick and dirty solution is to fetch the post inside `getServerSideProps`, or if possible pre-render the data, with `getStaticProps`, while still fetching fresh data client side, but with ISR revalidation that step feels a bit unnecessary, in some scenarios.
> 
> Basically:
> 
> * If you want the latest data at all times, no matter what, `GSSP` is your friend
> * If it can be stale first, and then (on refresh or after some time) be revalidated, `GSP` or `ISR`.
> 
> I am a bit intrigued that `curl` did show the UI though, maybe a misunderstanding.

Sorry, my bad. The curl showed the meta tags, but not the page in its loaded state. My bad. So it went down the not-loaded route.

**Reply by tavyy** [reply]
But I think I get what you mean. So basically FB will load the first rendered page which is the one in its not-loaded state. And that's why it never picks the updated meta tags. Damn.

**Reply by icyJoseph** [reply]
Yeah, pretty much that's what the Server Components, and the new suspense architecture are trying to solve in a more user friendly way, without blocking until everything is ready in one or another way.  Here's a great read on the topic https://github.com/reactwg/react-18/discussions/37.

Btw, this whole thing derives from react's render to string behaviour. You can experiment and learn a bit more of what's going on under the hood via, a codesandbox using CRA, https://codesandbox.io/s/new:

```js
// index.js
import { renderToString } from "react-dom/server";
import App from "./App";

const rootElement = document.getElementById("root");

const dehydrated = renderToString(<App />);

function _$(html, target) {
  // FOR DEMO ONLY DO NOT USE IN PROD
  let dom = new DOMParser().parseFromString(html, "text/html")..body; // adding a bug here to prevent copy pasting
  while (dom.hasChildNodes()) target.appendChild(dom.firstChilld); // maybe there's another bug 😉 ?
}

_$(dehydrated, rootElement);
```

And with `App` something like:

```js
import { useEffect, useState } from "react";

export default function App() {
  const [ready, setReady] = useState(false);

  useEffect(() => {
    const controller = new AbortController();

    fetch("https://pokeapi.co/api/v2/pokemon/ditto", {
      signal: controller.signal
    })
      .then((res) => res.json())
      .then(() => setReady(true));

    return () => controller.abort();
  }, []);

  if (!ready) return <div>Not ready</div>;

  return <div>Ready</div>;
}
```

Note that I added a couple of syntax errors in `_$`, to prevent people from just copy pasting blindly.

Anyway, you'll see that the server string doesn't contain the ready state, ever. You'll also see that no request leaves for the Pokemon API, because code inside effects is not taken into account during dehydration.
